File: services/auth/application.py
"""Application layer: команды и хендлеры для логина и регистрации."""
from __future__ import annotations

from dataclasses import dataclass

from urich.ddd import Command

from .infrastructure import AuthService


@dataclass
class Login(Command):
    username: str
    password: str


@dataclass
class Register(Command):
    username: str
    password: str
    role: str = "user"


class LoginHandler:
    def __init__(self, auth_service: AuthService) -> None:
        self._auth = auth_service

    async def __call__(self, cmd: Login) -> tuple[int, dict]:
        user = await self._auth.verify_user(cmd.username, cmd.password)
        if user is None:
            return 401, {"detail": "Invalid username or password"}
        return 200, {"user": user}  # токен добавляется в эндпоинте (нужен JWT_SECRET)


class RegisterHandler:
    def __init__(self, auth_service: AuthService) -> None:
        self._auth = auth_service

    async def __call__(self, cmd: Register) -> tuple[int, dict]:
        username = (cmd.username or "").strip()
        password = cmd.password or ""
        role = (cmd.role or "user").strip()
        if not username or not password:
            return 400, {"detail": "username and password required"}
        existing = await self._auth.get_by_username(username)
        if existing is not None:
            return 400, {"detail": "Username already exists"}
        import uuid
        user_id = str(uuid.uuid4())
        await self._auth.create_user(user_id, username, password, role)
        return 201, {"id": user_id, "username": username, "role": role}

File: rust/src/auth/application.rs
//! Обработчики login и register: Db из контейнера (DI, как services/auth). Свои ошибки — через .map_err(IntoCoreError::into_core_error).

use serde_json::{json, Value};
use urich_rs::{CoreError, IntoCoreError};
use uuid::Uuid;

use crate::auth::{Login, Register};
use crate::shared::Db;

fn hash_password(password: &str) -> String {
    format!("hash:{}", password.len())
}

fn verify_password(password: &str, hash: &str) -> bool {
    hash_password(password) == hash
}

pub async fn login_handler(cmd: Login, db: &Db) -> Result<Value, CoreError> {
    let username = cmd.username;
    let password = cmd.password;
    db.run(move |conn| {
        let mut stmt = conn
            .prepare("SELECT id, username, password_hash, role FROM users WHERE username = ?1")
            .map_err(IntoCoreError::into_core_error)?;
        let mut rows = stmt
            .query_map([&username], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, String>(2)?,
                    row.get::<_, String>(3)?,
                ))
            })
            .map_err(IntoCoreError::into_core_error)?;
        let row = rows.next();
        Ok(match row {
            None => json!({ "detail": "Invalid username or password" }),
            Some(Ok((id, username, password_hash, role))) if verify_password(&password, &password_hash) => {
                json!({
                    "user": { "id": id, "username": username, "role": role },
                    "token": format!("demo-token-{}", id)
                })
            }
            _ => json!({ "detail": "Invalid username or password" }),
        })
    })
    .await
}

pub async fn register_handler(cmd: Register, db: &Db) -> Result<Value, CoreError> {
    let username = cmd.username.trim().to_string();
    let password = cmd.password;
    let role = cmd.role.trim().to_string();
    if username.is_empty() || password.is_empty() {
        return Ok(json!({ "detail": "username and password required" }));
    }
    db.run(move |conn| {
        let exists: i64 = conn
            .query_row("SELECT COUNT(1) FROM users WHERE username = ?1", [&username], |r| r.get(0))
            .map_err(IntoCoreError::into_core_error)?;
        if exists > 0 {
            return Ok(json!({ "detail": "Username already exists" }));
        }
        let id = Uuid::new_v4().to_string();
        conn.execute(
            "INSERT INTO users (id, username, password_hash, role) VALUES (?1, ?2, ?3, ?4)",
            rusqlite::params![id, username.clone(), hash_password(&password), role.clone()],
        )
        .map_err(IntoCoreError::into_core_error)?;
        Ok(json!({ "id": id, "username": username, "role": role }))
    })
    .await
}
